cfg/posix.cfg

```diff
@@ -5456,6 +5461,7 @@ The function 'mktemp' is considered to be dangerous due to race conditions and s
   <!-- see http://man7.org/linux/man-pages/man3/strnlen.3.html-->
   <!-- size_t strnlen(const char *s, size_t maxlen); -->
   <function name="strnlen,wcsnlen">
+    <pure/>
```

Comment:
This is marked `pure` in the standard header but I do not understand how it could be.

Consider this (horrible) example:
```cpp
int len = strlen(p);
while (len > 0)
{
     p[len-5] = '\0';
     len = strlen(p);
}
```

The result of `strlen()` on the pointer would obviously be different on each call. So if the compiler would eliminate multiple calls to it we would get unexpected results.

Reply:
`strlen()` is designated as a pure function here: https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html

Reply:
The compiler somehow seems to know that it cannot be called less: https://godbolt.org/z/dKnfPreh7.

So if there is some additional logic involved within the compiler this attribute seems worthless (apart from implying `nodiscard`) in cases where we could apply that (like a possibly `variableScope` extension).

There is a proposal for `[[pure]]]`:
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0078r0.pdf. That seems to get into the case where you pass mutable data into a pure function.

Reply:
I assume the compiler might check if the input is somehow modified and leverages that information to make the decision.

Reading that paper I think it might sense to define our `pure` in the way section `6 Redefining pure` does and even create a check for this.

I will also file a clang-tidy ticket about this upstream.